src/v/kafka/server/quota_manager.cc

```diff
@@ -36,55 +55,114 @@ quota_manager::quota_manager()
-    return ss::make_ready_future<>();
+    co_return co_await ss::make_ready_future<>();
 }
 
-ss::future<quota_manager::client_quotas_t::iterator>
+ss::future<quota_manager::client_quotas_map_t::iterator>
 quota_manager::maybe_add_and_retrieve_quota(
   const std::optional<std::string_view>& quota_id,
   const clock::time_point& now) {
     // requests without a client id are grouped into an anonymous group that
     // shares a default quota. the anonymous group is keyed on empty string.
-    auto qid = quota_id ? *quota_id : "";
-
-    // find or create the throughput tracker for this client
-    //
-    // c++20: heterogeneous lookup for unordered_map can avoid creation of
-    // an sstring here but std::unordered_map::find isn't using an
-    // equal_to<> overload. this is a general issue we'll be looking at. for
-    // now, these client-name strings are small. This will be solved in
-    // c++20 via Hash::transparent_key_equal.
-    auto [it, inserted] = _client_quotas.try_emplace(
-      ss::sstring(qid),
-      client_quota{
-        now,
-        clock::duration(0),
-        {static_cast<size_t>(_default_num_windows()), _default_window_width()},
-        {static_cast<size_t>(_default_num_windows()), _default_window_width()},
-        /// pm_rate is only non-nullopt on the qm home shard
-        (ss::this_shard_id() == quota_manager_shard)
-          ? std::optional<token_bucket_rate_tracker>(
-            {*_target_partition_mutation_quota(),
-             static_cast<uint32_t>(_default_num_windows()),
-             _default_window_width()})
-          : std::optional<token_bucket_rate_tracker>()});
+    auto qid = ss::sstring{quota_id ? *quota_id : ""};
+
+    vassert(_client_quotas, "_client_quotas should have been initialized");
+
+    auto it = _client_quotas->find(qid);
+    auto inserted = false;
+    if (it == _client_quotas->end()) {
+        auto produce_rate = get_client_target_produce_tp_rate(qid);
+        auto fetch_rate = get_client_target_fetch_tp_rate(qid);
+        auto partition_mutation_rate = *_target_partition_mutation_quota();
+        auto default_num_windows = static_cast<uint32_t>(
+          _default_num_windows());
+        auto default_window_width = _default_window_width();
+
+        auto update_func =
+          [produce_rate,
+           fetch_rate,
+           partition_mutation_rate,
+           default_num_windows,
+           default_window_width,
+           qid,
+           now](const client_quotas_map_t& old_map) -> client_quotas_map_t {
+            auto new_map = client_quotas_map_t{old_map};
+
+            // Set the burst limit to the full window length to be similar to
+            // allow a similar level of bursting as Kafka's window-based
+            // algorithm
+            auto burst_limit_multiplier = default_num_windows
+                                          * default_window_width
+                                          / std::chrono::seconds(1);
```

Comment:
I don't think this bursting is necessary as the bucket is effectively limitless and tokens can always be taken.

Reply:
Tokens can be taken, but if the balance goes negative we start applying a delay -- the bursting prevents that and lets clients burst without delay.

I'm in two minds about whether that is useful or not, but that is the current behaviour and some tests start failing if we don't allow bursting because they depend on the bursting behaviour. But the tests only implicitly depend on this behaviour, specifically where they are producing 10 messages and assume that that's below the limit, whereas the request size of that is just above the limit, and only bursting allows those produce requests not to be throttled. So I think we could just reduce that 10 to ~8 if we don't want to allow bursting. Eg. here: https://github.com/redpanda-data/redpanda/blob/e89ed704260addd2f5398de874b2d9e134a4a9e7/tests/rptest/tests/cluster_quota_test.py#L192-L221